core/src/apps/bitcoin/sign_tx/bitcoin.py

```diff
@@ -150,6 +150,11 @@ async def step2_approve_outputs(self) -> None:
                 orig_txo = await self.get_original_output(txo, script_pubkey)
             await self.approve_output(txo, script_pubkey, orig_txo)
 
+        # Approve payment requests.
+        for index in self.approver.payment_requests:
```

Comment:
> Remind me, how important is it to allow interleaving of payment request outputs? It brings a lot of complications.

The only application that I had in mind was WabiSabi. The way I understand it, this would be a CoinJoin-like transaction which would be used to pay to another party. In order to reduce linkability of the payment input and payment output, the recipient might need to create multiple outputs, all of which need to be unlinkable to an outside observer (and even to the coordinator), so they shouldn't be clumped one after another, but randomly mixed among the others. For the user who is signing the transaction it would be tricky to verify the outputs manually and it would be much nicer to hide them under a payment request. But we should also consider that since the whole scheme is CoinJoin-like, the payment requests will probably need to be preauthorized, now that I am thinking about it.

So this capability seems to be quite far at the moment. I don't have anything against simplifying the current implementation as long as the protocol allows us to add this in the future in a backwards-compatible way.

> One more comment: verifying locality of the LTC/Dash/Doge addresses is nice and all, but it only works for Bitcoin-like coins, which seems limiting.

Yeah, I am not sure how to deal with that elegantly. BTW, the payment request memos contain a scriptPubKey, which seemed like a more generic solution to me than an address. Any thoughts on that?

Reply:
Let's first discuss the overall flow before we get to this. Regarding point (a) I think it's more error-prone to add another variable. It violates the single-point-of-truth principle.

On a side note, because of the reviewing of output addresses, I am a bit worried we might need to implement (c). Another alternative would be to return the amount field to `TxAckPaymentRequest` and allow only one pass.

Reply:
> Another alternative would be to return the amount field to `TxAckPaymentRequest` and allow only one pass.

how would that help? the current code is also one-pass

Reply:
...ah, right, but then we can pre-load the payment request instead of post-load

Reply:
> how would that help? the current code is also one-pass

I was referring to the reviewing of output addresses. You can either preload the payment request and then allow the user to review the individual outputs or you can postload it and perhaps even allow the user to cycle through them repeatedly (like the XPUBs in GetAddress). The second option is a littlebit tricky implementation-wise, because you either have to keep them in memory or stream all of them to recompute the `hash_outputs` to make sure an attacker is not giving you a fake one.